model-mesh-ovms-adapter/server/modelmanager.go

```diff
@@ -0,0 +1,504 @@
+	req.c <- status.Error(code, reason)
+}
+
+func (mm *OvmsModelManager) getConfig(ctx context.Context) error {
+	// query the Config Status API
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, fmt.Sprintf("%s/v1/config", mm.address), http.NoBody)
+	if err != nil {
+		// this should never happen...
+		return fmt.Errorf("Error constructing config status request: %w", err)
+	}
+
+	resp, err := mm.client.Do(req)
+	if err != nil {
+		// TODO: check if error is a timeout and handle appropriately
+		return fmt.Errorf("Protocol error getting the config: %w", err)
+	}
+	defer resp.Body.Close()
+
+	// read the body
+	// NOTE: if the body is not read, the connection cannot be re-used, so
+	// we read the body regardless of the status of the response
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return fmt.Errorf("Error reading config status response body: %w", err)
+	}
+
+	// handle successful request
+	if resp.StatusCode == http.StatusOK {
+		var c OvmsConfigResponse
+		if err1 := json.Unmarshal(body, &c); err1 != nil {
+			return fmt.Errorf("Error parsing config status response: %w", err1)
+		}
+		mm.cachedModelConfigResponse = c
+
+		return nil
+	}
+
+	var errorResponse OvmsConfigErrorResponse
+	if err = json.Unmarshal(body, &errorResponse); err != nil {
+		return fmt.Errorf("Error parsing model config error response: %w", err)
+	}
+
+	errDesc := fmt.Errorf("Error response when getting the config: %s", errorResponse.Error)
+	mm.log.Error(errDesc, "Call to /v1/config returned an error", "code", resp.StatusCode)
+
+	return status.Error(codes.Internal, errDesc.Error())
+}
+
+func (mm *OvmsModelManager) writeConfig() error {
+	// Build the model reposritory config to be written out
+	modelRepositoryConfig := OvmsMultiModelRepositoryConfig{
+		ModelConfigList: make([]OvmsMultiModelConfigListEntry, len(mm.loadedModelsMap)),
+	}
+	listIndex := 0
+	for _, model := range mm.loadedModelsMap {
+		modelRepositoryConfig.ModelConfigList[listIndex] = model
+		listIndex++
+	}
+
+	modelRepositoryConfigJSON, err := json.Marshal(modelRepositoryConfig)
+	if err != nil {
+		return fmt.Errorf("Error marshalling config file: %w", err)
+	}
+
+	if err := ioutil.WriteFile(mm.modelConfigFilename, modelRepositoryConfigJSON, 0644); err != nil {
+		return fmt.Errorf("Error writing config file: %w", err)
+	}
+
+	return nil
+}
+
+// updateModelConfig updates the model configuration for OVMS
+//
+// An error is returned if the reload was not confirmed to be completed; the
+// error will be nil even if a model load fails.
+//
+// The retruned config is saved to cachedModelConfigResponse.
+func (mm *OvmsModelManager) updateModelConfig(ctx context.Context) error {
+	if err := mm.writeConfig(); err != nil {
+		return fmt.Errorf("Error updating model config when writing config file: %w", err)
+	}
+
+	// Send config reload request to OVMS
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, fmt.Sprintf("%s/v1/config/reload", mm.address), http.NoBody)
```

Comment:
I'll move the endpoint strings to the struct to reduce calls to `fmt.Sprintf`.

For the timeout, there is a timeout enforced in `mm.client.Do()` used below. Do you think there should be another timeout added to the context separate from the HTTP timeout?

Reply:
Ah interesting, I had assumed this would only be controlled by the context. In that case, can we just store the request itself in the struct rather than the strings? I think it should be [ok to reuse](https://github.com/golang/go/issues/19653).

Reply:
I thought that they were controlling two different things. The timeout in the client limits how long an http request can take if nothing else happens with the context. The context included with the request allows the caller of `getConfig` to control that timeout or to cancel the request for another reason. Though, from [some searching](https://stackoverflow.com/questions/64129364/go-http-client-timeout-vs-context-timeout) it looks like they aren't really meant to be used together. Timeout in the client came before contexts were widely used and it seems that the context is recommended.

On the topic of re-use, I don't think we can reuse the context part of the request. The discussion you linked to is not using contexts in the requests. Though I did see that there is a `https://pkg.go.dev/net/http#Request.Clone` function to clone a request with a new context. I'm unsure if that saves us much on the request creation though. There is also a shallow copy via https://pkg.go.dev/net/http#Request.WithContext which may work, assuming that the context is the only field needs to change.

Alternatively, we could probably reuse the request if we don't use contexts and just rely on the timeout in the Transport (as in the discussion you linked). The trade-off there being that then caller does not have control over the timeout (which is probably fine here).

How do you want to proceed?